eth/tracers/api.go

```diff
@@ -725,139 +735,217 @@ txloop:
-			continue
-		}
-		// The transaction should be traced.
-		// Generate a unique temporary file to dump it into.
-		prefix := fmt.Sprintf("block_%#x-%d-%#x-", block.Hash().Bytes()[:4], i, tx.Hash().Bytes()[:4])
-		if !canon {
-			prefix = fmt.Sprintf("%valt-", prefix)
+	var signer = types.MakeSigner(api.backend.ChainConfig(), block.Number(), block.Time())
+	if config != nil && config.TxHash != (common.Hash{}) {
+		idx, ok := indexOf(block, config.TxHash)
+		if !ok {
+			return nil, fmt.Errorf("transaction %#x not found in block", config.TxHash)
 		}
-		var dump *os.File
-		dump, err := os.CreateTemp(os.TempDir(), prefix)
+		dump, err := api.standardTraceTxToFile(ctx, block, block.Transactions()[idx], idx, &config.Config, signer, canon)
 		if err != nil {
 			return nil, err
 		}
-		dumps = append(dumps, dump.Name())
-		// Set up the tracer and EVM for the transaction.
-		var (
-			writer = bufio.NewWriter(dump)
-			tracer = logger.NewJSONLogger(&logConfig, writer)
-			evm    = vm.NewEVM(vmctx, statedb, chainConfig, vm.Config{
-				Tracer:    tracer,
-				NoBaseFee: true,
-			})
-		)
-		// Execute the transaction and flush any traces to disk
-		statedb.SetTxContext(tx.Hash(), i)
-		if tracer.OnTxStart != nil {
-			tracer.OnTxStart(evm.GetVMContext(), tx, msg.From)
-		}
-		_, err = core.ApplyMessage(evm, msg, new(core.GasPool).AddGas(msg.GasLimit))
+		return []string{dump}, nil
+	}
+	// Retrieve the tracing configurations, or use default values
+	var logConfig logger.Config
+	if config != nil {
+		logConfig = config.Config
+	}
+
+	var (
+		dumps  []string
+		dump   *os.File
+		writer *bufio.Writer
+	)
+	instantiateTracer := func(ctx *Context) (*Tracer, error) {
+		// cleanup previous tx trace if one just ran
 		if writer != nil {
 			writer.Flush()
 		}
 		if dump != nil {
 			dump.Close()
 			log.Info("Wrote standard trace", "file", dump.Name())
 		}
+
+		// instantiate a new logger for this tx which will dump the trace to a temp file
+		dump, err := createTraceDumpFile(ctx.BlockHash, ctx.TxHash, ctx.TxIndex, canon)
 		if err != nil {
-			return dumps, err
+			return nil, err
 		}
-		// Finalize the state so any modifications are written to the trie
-		// Only delete empty objects if EIP158/161 (a.k.a Spurious Dragon) is in effect
-		statedb.Finalise(evm.ChainConfig().IsEIP158(block.Number()))
+		dumps = append(dumps, dump.Name())
 
-		// If we've traced the transaction we were looking for, abort
-		if tx.Hash() == txHash {
-			break
+		// Swap out the noop logger to the standard tracer
+		writer = bufio.NewWriter(dump)
+		tracer := Tracer{
+			logger.NewJSONLogger(&logConfig, writer),
+			func() (json.RawMessage, error) {
+				return nil, nil
+			},
+			func(err error) {},
 		}
+		return &tracer, nil
+	}
+
+	execOpt := standardTraceExecOpt(config)
+	execOpt.txTimeout = 100 * time.Hour
+	if _, err := api.traceBlock(ctx, block, execOpt, chainConfig, instantiateTracer); err != nil {
+		return nil, err
```

Comment:
If appear error, we also need to return the other successful traces.

Reply:
Good catch!  `traceBlock` will not emit the traces if an error occurs midway through the trace.

Reply:
So return `dumps[:len(dumps)-1], err` if len(dumps) > 1.